### ISSUE TITLE
Outgoing payments list problems (invoices don't work well with address-reuse)

### ISSUE THREAD

**Issue by snoopcatt** [spesmilo/electrum#6609]
Hello.

**Versions**
Version 4.0.3

**Summary**
When you making new transaction to address, that you have already sent bitcoins, new transaction instantly became «Paid» even if you cancelled transaction on confirmation/fee adjustment windw.

**Steps to reproduce**
1. Open «Send» tab
2. Send small amount of bitcoins (for example, 0.0001) to some address
3. Wait it to get at least 1 confirmation
4. Open «Send» tab agin
5. Try to create payment to same address:
5a. If you will try to send the same amount of bitcoins, as in step 2*, new «Invoice» won't appear at all.
5b. If you will try to send lesser amount of bitcoins, than in step 2* (for example, 0.00005), new «Invoice» will instantly get «Paid» status, even if you cancel transaction.
5c. If you will try to send greater amount of bitcoins, than in step 2* (for example, 0.0002), new «Invoice» will be created correctly.

* NB! If you have more than one payment to the same address, then you must count all related invoices, not just last.

**Expected behaviour**
Newly created invoices should stay unpaid before they will get a valid *txid* (after successfully broadcasted signed transaction).

**Patch**
I've made a «dirty» patch in order to fix this annoying behaviour:

```
--- electrum/gui/qt/main_window.py.orig	2020-09-21 22:47:49.979822918 +0300
+++ electrum/gui/qt/main_window.py	2020-09-21 23:14:33.820484941 +0300
@@ -1582,8 +1582,9 @@
             return
         self.wallet.save_invoice(invoice)
         self.invoice_list.update()
+        self.invoice_list.update_item(invoice.id, invoice, 0)
         self.do_clear()
-        self.do_pay_invoice(invoice)
+        self.do_pay_invoice(invoice) or self.delete_invoices([invoice.id])
 
     def pay_multiple_invoices(self, invoices):
         outputs = []
@@ -1659,6 +1660,7 @@
         else:
             self.preview_tx_dialog(make_tx=make_tx,
                                    external_keypairs=external_keypairs)
+        return True
 
     def preview_tx_dialog(self, *, make_tx, external_keypairs=None):
         d = PreviewTxDialog(make_tx=make_tx, external_keypairs=external_keypairs,
```

```
--- electrum/gui/qt/invoice_list.py.orig	2020-09-21 22:43:04.740094826 +0300
+++ electrum/gui/qt/invoice_list.py	2020-09-21 22:42:45.859549228 +0300
@@ -75,7 +75,7 @@
         self.setSelectionMode(QAbstractItemView.ExtendedSelection)
         self.update()
 
-    def update_item(self, key, invoice: Invoice):
+    def update_item(self, key, invoice: Invoice, status=None):
         model = self.std_model
         for row in range(0, model.rowCount()):
             item = model.item(row, 0)
@@ -84,7 +84,7 @@
         else:
             return
         status_item = model.item(row, self.Columns.STATUS)
-        status = self.parent.wallet.get_invoice_status(invoice)
+        status = (status == None) and self.parent.wallet.get_invoice_status(invoice) or status 
         status_str = invoice.get_status_str(status)
         if self.parent.wallet.lnworker:
             log = self.parent.wallet.lnworker.logs.get(key)
```

But it is wrong way. I suggest to fix this issue by more precious way. 
I imagine that way:

1. Add `txid` field to *invoice* table in wallet database. `wallet_db.py`
2. Treat all on-chain invoices without `txid` as unpaid.  `wallet.py, get_invoice_status()`
3. After transaction successfully signed and broadcasted, appear `txid` to invoice. `main_window.py, pay_onchain_dialog()`

But I don't know Electrum internal architecture and I have no idea how to update database structure in correct way.
I hope I have clearly explained and will help fix this annoying bug.

Thank you!

**Comment by blackhand23** [comment]
My payment still has not gone through and my transaction has 112 confirmations.

**Comment by snoopcatt** [comment]
> My payment still has not gone through and my transaction has 112 confirmations.

I don't think it is related to this issue.

**Comment by SomberNight** [comment]
In short, the issue is that invoices don't work well with address-reuse.

We let invoices be satisfied by both "past" and "future" transactions.
Invoices can be imported/exported. When imported, they should become "paid" if a "past" tx satisfies them.
Also, a tx can satisfy multiple invoices, and multiple txs added up can satisfy a single invoice.

Also, one could argue that a single 1 BTC tx output to address A should not satisfy two invoices asking for 1 BTC at address A, but which invoice should be marked "paid" then?

**Comment by SomberNight** [comment]
note: we could store a block height in each invoice, and `is_paid` logic could consider only blocks with a greater height

**Comment by ecdsa** [comment]
this needs to be fixed for #6824
